File: src/trck/diff.py
from __future__ import annotations
from dataclasses import dataclass, field
from pathlib import Path
import sys
from .config import status_names
from .constants import ITEMS_DIR, die
from .index import CANON_KEYS, Ctx, Issue, filename, parse_index

# --------------------------------------------------------------------------- #
# diff: the source seam
# --------------------------------------------------------------------------- #
# `trck diff` compares the tracker at two points, and it must not require git.
# Every query and mutate verb is VCS-free; only `install-hook` and `setup-git`
# shell out, and both are explicitly git-flavoured. So the dependency is isolated
# here rather than assumed: everything downstream consumes a `Snapshot` and cannot
# tell where it came from. Git is one provider among several, added on top.
TIMESTAMP_FIELDS = ("created", "started", "closed")
SET_FIELDS = ("labels", "depends_on")
# Compared field-by-field: every known field except the id (the join key) and the
# timestamps (which are the *evidence* for a status change, not a change of their
# own). Set-valued fields are compared as sets, separately.
SCALAR_FIELDS = tuple(k for k in CANON_KEYS
                      if k != "id" and k not in TIMESTAMP_FIELDS and k not in SET_FIELDS)


class Snapshot:
    """The tracker's state at one point, plus a label for display.

    `body_reader` maps a row to its markdown text, or is None when the source
    cannot supply bodies at all (a bare index.jsonl, stdin). That distinction is
    load-bearing: `body()` returning None means "unavailable from this source",
    which is NOT the same as "" meaning "the body is empty" — reporting the first
    as the second would silently under-report body edits.

    A reader, rather than a directory, is what lets a provider that has no
    directory to point at (a git revision, say) supply bodies on the same terms.
    Bodies are read lazily, so a diff that never asks for one costs nothing.
    """
    def __init__(self, rows: list[Issue], label: str, body_reader=None):
        self.rows = rows
        self.label = label
        self._read_body = body_reader
        self._by_id = {r.id: r for r in rows}

    @property
    def has_bodies(self) -> bool:
        return self._read_body is not None

    def row(self, iid: str) -> Issue | None:
        return self._by_id.get(iid)

    def body(self, iid: str) -> str | None:
        """The issue's markdown body, or None when unavailable — either because the
        source has no bodies at all, or because this issue has none in it."""
        row = self._by_id.get(iid)
        if self._read_body is None or row is None:
            return None
        return self._read_body(row)


def dir_body_reader(items: Path):
    """Read bodies out of an on-disk `items/` directory."""
    def read(row: Issue) -> str | None:
        p = items / filename(row)
        return p.read_text() if p.exists() else None
    return read


def snapshot_from_text(text: str, label: str) -> Snapshot:
    """A rows-only snapshot from index.jsonl text (stdin, or a blob read by a provider)."""
    return Snapshot(parse_index(text, label), label)


def snapshot_from_index(path: Path, label: str | None = None) -> Snapshot:
    """A rows-only snapshot from an index.jsonl file: no bodies alongside it."""
    path = Path(path)
    label = label or path.name
    if not path.exists():
        die(f"no such file: {path}")
    return Snapshot(parse_index(path.read_text(), label), label)


def snapshot_from_dir(path: Path, label: str | None = None) -> Snapshot:
    """A snapshot from a whole tracker dir — rows and bodies both.

    A dir with no index.jsonl is an empty snapshot, not an error: the tracker not
    existing on one side is a legitimate comparison (everything reads as added).
    """
    path = Path(path)
    label = label or path.name
    index = path / "index.jsonl"
    rows = parse_index(index.read_text(), label) if index.exists() else []
    return Snapshot(rows, label, body_reader=dir_body_reader(path / ITEMS_DIR))


def snapshot_working_tree(ctx: Ctx) -> Snapshot:
    return snapshot_from_dir(ctx.dir, "working tree")


def resolve_source(spec: str | None, ctx: Ctx | None = None) -> Snapshot:
    """Turn a `--from`/`--to` value into a Snapshot.

    `-` reads index.jsonl from stdin, a directory is read as a tracker dir (bodies
    included), any other path as a bare index.jsonl. None means the working tree,
    which needs a resolved `ctx`.
    """
    if spec is None:
        if ctx is None:
            die("no tracker to compare against")
        return snapshot_working_tree(ctx)
    if spec == "-":
        return snapshot_from_text(sys.stdin.read(), "stdin")
    path = Path(spec)
    if path.is_dir():
        return snapshot_from_dir(path)
    return snapshot_from_index(path)


# --------------------------------------------------------------------------- #
# diff: the change model
# --------------------------------------------------------------------------- #
@dataclass
class FieldDelta:
    """A scalar field that differs between the two sides."""
    name: str
    old: object
    new: object


@dataclass
class SetDelta:
    """A set-valued field (labels, depends_on) with its gained and lost members."""
    name: str
    added: list
    removed: list


@dataclass
class Change:
    """What happened to one issue between the two snapshots.

    `kind` is added / removed / modified. `direction` describes a status move
    against the *configured* status order — forward, backward, or lateral when
    either side's status is outside the current vocabulary — and is None when the
    status did not move. Renderers need it because a `done -> ongoing` reopen must
    not read like a `backlog -> ongoing` start.
    """
    id: str
    kind: str
    old: Issue | None
    new: Issue | None
    fields: list[FieldDelta] = field(default_factory=list)
    sets: list[SetDelta] = field(default_factory=list)
    timestamps: dict = field(default_factory=dict)
    direction: str | None = None


@dataclass
class Diff:
    """The change records plus both snapshots — renderers need the full rows for
    titles, icons, and rollups, not just what changed."""
    old: Snapshot
    new: Snapshot
    changes: list[Change]


def status_direction(cfg: dict, old: str, new: str) -> str | None:
    """Classify a status move against the configured order. A status the current
    trck.json doesn't know (an old snapshot written under a renamed vocabulary) is
    unordered, so the move is 'lateral' rather than an error."""
    if old == new:
        return None
    order = status_names(cfg)
    if old not in order or new not in order:
        return "lateral"
    return "forward" if order.index(new) > order.index(old) else "backward"


def _values(row: Issue) -> dict:
    """Every comparable scalar of a row, built-in and custom alike."""
    full = row.to_dict()
    return {k: full.get(k) for k in SCALAR_FIELDS} | dict(row.extra)


def _compare(cfg: dict, old: Issue, new: Issue) -> Change | None:
    """One issue present on both sides: None when nothing moved."""
    ov, nv = _values(old), _values(new)
    fields = [FieldDelta(k, ov.get(k), nv.get(k))
              for k in sorted(ov.keys() | nv.keys()) if ov.get(k) != nv.get(k)]
    sets = []
    for k in SET_FIELDS:
        a, b = set(getattr(old, k) or []), set(getattr(new, k) or [])
        if a != b:
            sets.append(SetDelta(k, sorted(b - a), sorted(a - b)))
    stamps = {k: (getattr(old, k), getattr(new, k))
              for k in TIMESTAMP_FIELDS if getattr(old, k) != getattr(new, k)}
    if not (fields or sets or stamps):
        return None
    return Change(id=new.id, kind="modified", old=old, new=new, fields=fields,
                  sets=sets, timestamps=stamps,
                  direction=status_direction(cfg, old.status, new.status))


def diff_snapshots(cfg: dict, old: Snapshot, new: Snapshot) -> Diff:
    """Join two snapshots by id and classify what changed. Pure: no I/O, no notion
    of a revision — whoever produced the snapshots owns that."""
    olds = {r.id: r for r in old.rows}
    news = {r.id: r for r in new.rows}
    changes = []
    for iid in sorted(olds.keys() | news.keys()):
        o, n = olds.get(iid), news.get(iid)
        if o is None:
            changes.append(Change(id=iid, kind="added", old=None, new=n))
        elif n is None:
            changes.append(Change(id=iid, kind="removed", old=o, new=None))
        elif (c := _compare(cfg, o, n)) is not None:
            changes.append(c)
    return Diff(old=old, new=new, changes=changes)

File: src/diff.rs
//! `diff` — what changed between two tracker states — and `changelog`.
//!
//! The design worth knowing is the **source seam**: `diff` compares two [`Snapshot`]s and
//! has no notion of a revision. Whoever produced the snapshots owns that, which is what
//! lets `--from FILE`, `--from -`, a git revision and the working tree all feed the same
//! comparison. Git is a layer on top, not a dependency of the model.
//!
//! The change model is deliberately three-way. A scalar field that moved, a set field
//! that gained and lost members, and a timestamp are different questions, and flattening
//! them into "these keys differ" would lose the one thing a reader wants — whether a
//! `done -> ongoing` move is a reopen or a start.

use crate::config::{self, is_terminal};
use crate::discovery::Ctx;
use crate::graph::Graph;
use crate::index::parse_index;
use crate::issue::{CANON_KEYS, Issue};
use crate::render::field_value;
use std::collections::{BTreeMap, BTreeSet};
use std::path::Path;

const TIMESTAMP_FIELDS: &[&str] = &["created", "started", "closed"];
const SET_FIELDS: &[&str] = &["labels", "depends_on"];

/// Everything comparable as a scalar: the canonical fields minus the id, the timestamps
/// and the set-valued ones.
fn scalar_fields() -> Vec<&'static str> {
    CANON_KEYS.iter().copied().filter(|k| *k != "id" && !TIMESTAMP_FIELDS.contains(k) && !SET_FIELDS.contains(k)).collect()
}

/// A tracker state, and a label naming where it came from.
pub(crate) struct Snapshot {
    pub(crate) label: String,
    pub(crate) rows: Vec<Issue>,
}

impl Snapshot {
    fn from_text(text: &str, label: &str) -> Result<Snapshot, String> {
        Ok(Snapshot { label: label.to_string(), rows: parse_index(text, label)? })
    }
}

/// A scalar field that differs between the two sides.
pub(crate) struct FieldDelta {
    pub(crate) name: String,
    pub(crate) old: Option<String>,
    pub(crate) new: Option<String>,
}

/// A set-valued field with what it gained and lost.
pub(crate) struct SetDelta {
    pub(crate) name: String,
    pub(crate) added: Vec<String>,
    pub(crate) removed: Vec<String>,
}

/// What happened to one issue between the two snapshots.
pub(crate) struct Change {
    pub(crate) id: String,
    pub(crate) kind: &'static str,
    pub(crate) title: String,
    pub(crate) fields: Vec<FieldDelta>,
    pub(crate) sets: Vec<SetDelta>,
    pub(crate) timestamps: BTreeMap<String, (Option<String>, Option<String>)>,
    /// `forward`, `backward` or `lateral`; `None` when the status did not move.
    pub(crate) direction: Option<&'static str>,
}

/// Classify a status move against the vocabulary order.
///
/// A status the engine does not know — an old snapshot written under a renamed
/// vocabulary — is unordered, so the move is `lateral` rather than an error. A renderer
/// needs this because a `done -> ongoing` reopen must not read like a `backlog ->
/// ongoing` start.
pub(crate) fn status_direction(old: &str, new: &str) -> Option<&'static str> {
    if old == new {
        return None;
    }
    let pos = |s: &str| config::STATUSES.iter().position(|x| *x == s);
    match (pos(old), pos(new)) {
        (Some(a), Some(b)) if b > a => Some("forward"),
        (Some(_), Some(_)) => Some("backward"),
        _ => Some("lateral"),
    }
}

/// Every comparable scalar of a row, built-in and custom alike.
fn values(row: &Issue) -> BTreeMap<String, String> {
    let mut out: BTreeMap<String, String> = scalar_fields().into_iter().filter_map(|k| field_value(row, k).map(|v| (k.to_string(), v))).collect();
    for (k, v) in &row.extra {
        if let Some(s) = v.as_str() {
            out.insert(k.clone(), s.to_string());
        } else {
            out.insert(k.clone(), v.to_json());
        }
    }
    out
}

/// One issue present on both sides. `None` when nothing moved.
fn compare(old: &Issue, new: &Issue) -> Option<Change> {
    let (ov, nv) = (values(old), values(new));
    let keys: BTreeSet<&String> = ov.keys().chain(nv.keys()).collect();
    let fields: Vec<FieldDelta> = keys
        .into_iter()
        .filter(|k| ov.get(*k) != nv.get(*k))
        .map(|k| FieldDelta { name: k.clone(), old: ov.get(k).cloned(), new: nv.get(k).cloned() })
        .collect();

    let mut sets = Vec::new();
    for name in SET_FIELDS {
        let pick =
            |r: &Issue| -> BTreeSet<String> { if *name == "labels" { r.labels.iter().cloned().collect() } else { r.depends_on.iter().cloned().collect() } };
        let (a, b) = (pick(old), pick(new));
        if a != b {
            sets.push(SetDelta { name: (*name).to_string(), added: b.difference(&a).cloned().collect(), removed: a.difference(&b).cloned().collect() });
        }
    }

    let stamp = |r: &Issue, name: &str| match name {
        "created" => r.created.clone(),
        "started" => r.started.clone(),
        _ => r.closed.clone(),
    };
    let timestamps: BTreeMap<String, (Option<String>, Option<String>)> =
        TIMESTAMP_FIELDS.iter().filter(|k| stamp(old, k) != stamp(new, k)).map(|k| ((*k).to_string(), (stamp(old, k), stamp(new, k)))).collect();

    if fields.is_empty() && sets.is_empty() && timestamps.is_empty() {
        return None;
    }
    Some(Change {
        id: new.id.clone(),
        kind: "modified",
        title: new.title.clone(),
        fields,
        sets,
        timestamps,
        direction: status_direction(&old.status, &new.status),
    })
}

/// Join two snapshots by id and classify what changed. Pure: no I/O, and no notion of a
/// revision — whoever produced the snapshots owns that.
pub(crate) fn diff_snapshots(old: &Snapshot, new: &Snapshot) -> Vec<Change> {
    let olds: BTreeMap<&str, &Issue> = old.rows.iter().map(|r| (r.id.as_str(), r)).collect();
    let news: BTreeMap<&str, &Issue> = new.rows.iter().map(|r| (r.id.as_str(), r)).collect();
    let ids: BTreeSet<&str> = olds.keys().chain(news.keys()).copied().collect();
    let mut changes = Vec::new();
    for id in ids {
        match (olds.get(id), news.get(id)) {
            (None, Some(n)) => changes.push(Change {
                id: id.to_string(),
                kind: "added",
                title: n.title.clone(),
                fields: Vec::new(),
                sets: Vec::new(),
                timestamps: BTreeMap::new(),
                direction: None,
            }),
            (Some(o), None) => changes.push(Change {
                id: id.to_string(),
                kind: "removed",
                title: o.title.clone(),
                fields: Vec::new(),
                sets: Vec::new(),
                timestamps: BTreeMap::new(),
                direction: None,
            }),
            (Some(o), Some(n)) => {
                if let Some(c) = compare(o, n) {
                    changes.push(c);
                }
            },
            (None, None) => {},
        }
    }
    changes
}

/// A compact, plain-text account of what moved on one issue.
pub(crate) fn change_summary(c: &Change) -> String {
    let show = |v: &Option<String>| v.clone().unwrap_or_else(|| "None".to_string());
    let mut bits: Vec<String> = c.fields.iter().map(|f| format!("{} {} → {}", f.name, show(&f.old), show(&f.new))).collect();
    for s in &c.sets {
        let members: Vec<String> = s.added.iter().map(|v| format!("+{v}")).chain(s.removed.iter().map(|v| format!("-{v}"))).collect();
        bits.push(format!("{} {}", s.name, members.join(" ")));
    }
    if bits.is_empty() {
        // A timestamp-only edit still changed something; say what.
        bits = c.timestamps.iter().map(|(k, (a, b))| format!("{k} {} → {}", show(a), show(b))).collect();
    }
    bits.join(", ")
}

// --------------------------------------------------------------------------- //
// sources
// --------------------------------------------------------------------------- //

const USE_FROM: &str = "use --from/--to with file paths instead";

fn git_run(args: &[&str], cwd: &Path) -> Result<std::process::Output, String> {
    std::process::Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .map_err(|_| format!("git is not on PATH, so revision specs are unavailable; {USE_FROM}"))
}

/// The tracker dir as a repo-relative prefix, the way `git show <rev>:<path>` wants it.
/// A tracker dir that *is* the repo root yields an empty prefix.
fn git_tracker_prefix(ctx: &Ctx) -> Result<String, String> {
    let out = git_run(&["rev-parse", "--show-toplevel"], &ctx.dir)?;
    if !out.status.success() {
        return Err(format!("not a git repository, so revision specs are unavailable; {USE_FROM}"));
    }
    let root = Path::new(String::from_utf8_lossy(&out.stdout).trim()).to_path_buf();
    let dir = ctx.dir.canonicalize().unwrap_or_else(|_| ctx.dir.clone());
    let root = root.canonicalize().unwrap_or(root);
    let rel = dir.strip_prefix(&root).map_err(|_| format!("tracker dir {} is not inside the git repo at {}", ctx.dir.display(), root.display()))?;
    let rel = rel.to_string_lossy().replace('\\', "/");
    Ok(if rel.is_empty() || rel == "." { String::new() } else { format!("{rel}/") })
}

/// The tracker as of `rev`.
///
/// A tracker dir absent at that revision is **not** an error: comparing against a commit
/// from before the tracker existed is a legitimate question, and the answer is "every
/// issue is new". An unresolvable revision *is* an error, reported separately, so "you
/// typo'd the branch" stays distinguishable from "the tracker did not exist yet".
pub(crate) fn git_snapshot(ctx: &Ctx, rev: &str) -> Result<Snapshot, String> {
    let verify = git_run(&["rev-parse", "--verify", "--quiet", &format!("{rev}^{{commit}}")], &ctx.dir)?;
    if !verify.status.success() {
        return Err(format!("unknown revision '{rev}'"));
    }
    let prefix = git_tracker_prefix(ctx)?;
    let path = format!("{prefix}index.jsonl");
    let out = git_run(&["show", &format!("{rev}:{path}")], &ctx.dir)?;
    let text = if out.status.success() { String::from_utf8_lossy(&out.stdout).into_owned() } else { String::new() };
    Snapshot::from_text(&text, rev)
}

/// Split a revision spec into `(old, new)`; a `None` new side means the working tree.
pub(crate) fn parse_rev_spec(spec: &str) -> Result<(String, Option<String>), String> {
    if spec.contains("...") {
        return Err("three-dot (merge-base) revision specs are not supported; \
                    use `a..b` to compare two revisions directly"
            .to_string());
    }
    let Some((old, new)) = spec.split_once("..") else {
        return Ok((spec.to_string(), None));
    };
    if old.is_empty() || new.is_empty() {
        return Err(format!("incomplete revision range '{spec}'; both sides of `..` are required"));
    }
    Ok((old.to_string(), Some(new.to_string())))
}

/// Resolve a `--from`/`--to` spec: a file, a directory holding one, `-` for stdin, or the
/// working tree when unspecified.
pub(crate) fn resolve_source(spec: Option<&str>, ctx: &Ctx) -> Result<Snapshot, String> {
    let Some(spec) = spec else {
        let text = std::fs::read_to_string(ctx.index_path()).unwrap_or_default();
        return Snapshot::from_text(&text, "working tree");
    };
    if spec == "-" {
        use std::io::Read as _;
        let mut text = String::new();
        std::io::stdin().read_to_string(&mut text).map_err(|e| format!("stdin: {e}"))?;
        return Snapshot::from_text(&text, "stdin");
    }
    let path = Path::new(spec);
    // The label is the file's own name, not the spec that named it: a long relative path
    // buries the one word that identifies the side being compared.
    let label = path.file_name().map_or_else(|| spec.to_string(), |n| n.to_string_lossy().into_owned());
    if path.is_dir() {
        // A tracker dir with no index is an empty snapshot, not an error: the tracker
        // not existing on one side is a legitimate comparison, and everything on the
        // other side reads as added.
        let text = std::fs::read_to_string(path.join("index.jsonl")).unwrap_or_default();
        return Snapshot::from_text(&text, &label);
    }
    let text = std::fs::read_to_string(path).map_err(|_| format!("no such file: {spec}"))?;
    Snapshot::from_text(&text, &label)
}

// --------------------------------------------------------------------------- //
// changelog
// --------------------------------------------------------------------------- //

/// Validate a `--since` cutoff: a bare date or a full UTC timestamp.
pub(crate) fn parse_since(value: &str) -> Result<String, String> {
    let date_ok = value.len() >= 10 && value.as_bytes()[..10].iter().enumerate().all(|(i, b)| if i == 4 || i == 7 { *b == b'-' } else { b.is_ascii_digit() });
    let ok = date_ok
        && (value.len() == 10
            || (value.len() == 20
                && value.as_bytes()[10] == b'T'
                && value.ends_with('Z')
                && value[11..19].bytes().enumerate().all(|(i, b)| if i == 2 || i == 5 { b == b':' } else { b.is_ascii_digit() })));
    if ok { Ok(value.to_string()) } else { Err(format!("--since must be a date (YYYY-MM-DD) or timestamp (YYYY-MM-DDTHH:MM:SSZ), got '{value}'")) }
}

/// Issues that *shipped* on or after `since`.
///
/// Terminal, closed in the window, and carrying **no resolution** — so wontfix, duplicate
/// and superseded are excluded. That absence is the whole rule: `done` alone does not
/// mean anything shipped, and an issue closed without producing something has no release
/// note to write.
pub(crate) fn select_shipped(rows: &[Issue], since: &str) -> Vec<Issue> {
    rows.iter()
        .filter(|r| is_terminal(&r.status))
        .filter(|r| r.closed.as_deref().is_some_and(|c| c >= since))
        .filter(|r| r.resolution.is_none())
        .cloned()
        .collect()
}

fn walk(g: &Graph, out: &mut Vec<String>, id: &str, depth: usize, seen: &BTreeSet<String>) {
    let Some(node) = g.get(id) else { return };
    let tag = node.extra.get("component").and_then(|v| v.as_str()).map_or_else(String::new, |c| format!(" ({c})"));
    out.push(format!("{}- #{} {}{tag}", "  ".repeat(depth), node.id, node.title));
    if seen.contains(id) {
        return;
    }
    let mut kids: Vec<String> = g.children_of(id).to_vec();
    kids.sort();
    kids.sort_by(|a, b| {
        let closed = |i: &String| g.get(i).and_then(|r| r.closed.clone()).unwrap_or_default();
        closed(b).cmp(&closed(a))
    });
    let mut next = seen.clone();
    next.insert(id.to_string());
    for kid in kids {
        walk(g, out, &kid, depth + 1, &next);
    }
}

/// Render the changelog: shipped issues nested under shipped parents, newest first.
pub(crate) fn render_changelog(shipped: &[Issue], since: &str) -> String {
    let n = shipped.len();
    let header = format!("## Shipped since {since} — {n} issue{}", if n == 1 { "" } else { "s" });
    if shipped.is_empty() {
        return format!("{header}\n\n_none_\n");
    }
    let g = Graph::new(shipped.to_vec());
    let mut out = vec![header, String::new()];

    // Closed descending, id ascending on ties — a stable sort over an id-sorted input.
    let sib_sorted = |mut ids: Vec<String>| -> Vec<String> {
        ids.sort();
        ids.sort_by(|a, b| {
            let closed = |i: &String| g.get(i).and_then(|r| r.closed.clone()).unwrap_or_default();
            closed(b).cmp(&closed(a))
        });
        ids
    };

    let roots: Vec<String> = g.rows.iter().filter(|r| r.parent.as_ref().is_none_or(|p| g.get(p).is_none())).map(|r| r.id.clone()).collect();
    for root in sib_sorted(roots) {
        walk(&g, &mut out, &root, 0, &BTreeSet::new());
    }
    out.push(String::new());
    out.join("\n")
}

#[cfg(test)]
mod tests {
    // Tests assert; that is their job. The crate denies unwrap/expect/panic because a
    // malformed tracker must produce a diagnostic rather than a stack trace, but a test
    // that cannot panic cannot fail.
    #![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic)]

    use super::*;
    use crate::json::parse;

    fn issue(json: &str) -> Issue {
        Issue::from_json(&parse(json).expect("json")).expect("issue")
    }

    const BASE: &str = r#"{"id": "aaaaaaa", "slug": "a", "title": "A", "status": "backlog", "priority": "medium"}"#;

    #[test]
    fn a_reopen_reads_differently_from_a_start() {
        // The reason `direction` exists: both are status moves, and a renderer that
        // cannot tell them apart reports a regression as progress.
        assert_eq!(status_direction("backlog", "ongoing"), Some("forward"));
        assert_eq!(status_direction("done", "ongoing"), Some("backward"));
        assert_eq!(status_direction("done", "done"), None);
    }

    #[test]
    fn a_status_outside_the_vocabulary_is_lateral_not_an_error() {
        // An old snapshot written under a renamed vocabulary is unordered, not invalid.
        assert_eq!(status_direction("shipped", "done"), Some("lateral"));
    }

    #[test]
    fn an_unchanged_issue_produces_no_change() {
        let a = issue(BASE);
        assert!(compare(&a, &a).is_none());
    }

    #[test]
    fn a_scalar_move_is_reported_with_both_sides() {
        let a = issue(BASE);
        let b = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "urgent""#));
        let c = compare(&a, &b).expect("changed");
        assert_eq!(c.fields.len(), 1);
        assert_eq!(c.fields[0].name, "priority");
        assert_eq!(c.fields[0].old.as_deref(), Some("medium"));
        assert_eq!(c.fields[0].new.as_deref(), Some("urgent"));
    }

    #[test]
    fn a_set_field_reports_what_it_gained_and_lost() {
        let a = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "medium", "labels": ["x", "y"]"#));
        let b = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "medium", "labels": ["y", "z"]"#));
        let c = compare(&a, &b).expect("changed");
        assert_eq!(c.sets.len(), 1);
        assert_eq!(c.sets[0].added, ["z"]);
        assert_eq!(c.sets[0].removed, ["x"]);
    }

    #[test]
    fn a_timestamp_only_edit_still_says_what_moved() {
        let a = issue(BASE);
        let b = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "medium", "started": "2026-01-01T00:00:00Z""#));
        let c = compare(&a, &b).expect("changed");
        assert!(c.fields.is_empty() && c.sets.is_empty());
        assert!(change_summary(&c).contains("started"));
    }

    #[test]
    fn a_custom_field_is_compared_like_any_other_scalar() {
        let a = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "medium", "assignee": "alice""#));
        let b = issue(&BASE.replace(r#""priority": "medium""#, r#""priority": "medium", "assignee": "bo""#));
        let c = compare(&a, &b).expect("changed");
        assert_eq!(c.fields[0].name, "assignee");
    }

    #[test]
    fn added_and_removed_are_joined_by_id() {
        let old = Snapshot { label: "old".into(), rows: vec![issue(BASE)] };
        let new = Snapshot { label: "new".into(), rows: vec![issue(&BASE.replace("aaaaaaa", "bbbbbbb"))] };
        let changes = diff_snapshots(&old, &new);
        // Joined by id, and reported in id order: aaaaaaa (gone) before bbbbbbb (new).
        let seen: Vec<(&str, &str)> = changes.iter().map(|c| (c.id.as_str(), c.kind)).collect();
        assert_eq!(seen, [("aaaaaaa", "removed"), ("bbbbbbb", "added")]);
    }

    #[test]
    fn a_revision_range_names_both_sides() {
        assert_eq!(parse_rev_spec("HEAD").expect("ok"), ("HEAD".into(), None));
        assert_eq!(parse_rev_spec("a..b").expect("ok"), ("a".to_string(), Some("b".to_string())));
        assert!(parse_rev_spec("a...b").is_err(), "merge-base specs are refused");
        assert!(parse_rev_spec("a..").is_err());
        assert!(parse_rev_spec("..b").is_err());
    }

    #[test]
    fn since_takes_a_date_or_a_full_timestamp() {
        assert!(parse_since("2026-01-01").is_ok());
        assert!(parse_since("2026-01-01T00:00:00Z").is_ok());
        for bad in ["2026", "01-01-2026", "2026-01-01T00:00:00", "yesterday", ""] {
            assert!(parse_since(bad).is_err(), "should reject {bad}");
        }
    }

    #[test]
    fn shipped_excludes_anything_closed_without_shipping() {
        // The absence of a resolution is the whole rule: `done` alone does not mean
        // something shipped.
        let mk = |id: &str, res: &str| {
            issue(&format!(
                r#"{{"id": "{id}", "slug": "s", "title": "T", "status": "done", "priority": "low", "closed": "2026-06-11T00:00:00Z"{}}}"#,
                if res.is_empty() { String::new() } else { format!(r#", "resolution": "{res}""#) }
            ))
        };
        let rows = vec![mk("aaaaaaa", ""), mk("bbbbbbb", "wontfix"), issue(BASE)];
        let shipped = select_shipped(&rows, "2026-06-01");
        assert_eq!(shipped.iter().map(|r| r.id.as_str()).collect::<Vec<_>>(), ["aaaaaaa"]);
    }

    #[test]
    fn an_empty_changelog_says_none() {
        let out = render_changelog(&[], "2026-01-01");
        assert!(out.contains("0 issues"), "{out}");
        assert!(out.ends_with("_none_\n"), "{out}");
    }
}
